### DISCUSSION TITLE
Exposing Midi App to Other Apps

### DISCUSSION THREAD

**Discussion by SNewtown** [orchetect/swift-midi#213] [Q&A]
Hi Steffan,  I am still using midikit and I can’t begin to express how great it is!  

I did have a question and it has to do with exposing my app (which includes midikit) to other apps so that other apps can connect to it.  

My app can currently send and receive midi messages just fine, but that entails that the other apps broadcast midi to all apps/devices (which many do) but some apps will only send to specifically connected apps.  

Could you point me to the documentation on what needs to be enabled for my instance of midikit to become visible as a connectable app for other apps?  

I am using the prior release of midikit, not the very latest…. Thanks!

**Comment by orchetect** [comment]
What platform? iOS?

I need more details on how it's currently set up. What MIDIKit API is being used to create the send / receive ports or connections?

**Reply by SNewtown** [reply]
Sorry, realizing I didn’t provide much context.  Definitely iOS and using midi kit from about 8 or 9 months ago and will provide more details once I am home later today as to the exact version and how the manager is setup.  

But the gist of it is that it is working perfectly well with devices connected and apps that send/receive midi without having to connect.  But some apps will only send to connected apps and when I list the available apps in those other apps I do not see mine - so I figured there was some switch or something to make my app with midikit visible to other apps

**Reply by SNewtown** [reply]
Hi Steffan,

I can confirm that I am using mIdikit 0.8.8 and my connection is added like this:

```swift
try midiManager.addInputConnection(
                toOutputs: [],
                tag: "Listener",
                mode: .allEndpoints,
                filter: .owned(), 
                receiver: .events{[self] events in
                    DispatchQueue.main.async {
                        events.forEach { midiListener.receivedMIDIEvent($0) }
                       }
                }
            )
```

and

```swift            
try midiManager.addOutputConnection(
                toInputs: [], 
                tag: "Broadcaster",
                mode: .allEndpoints,
                filter: .owned()
            )
```

Again, sending/receiving works perfectly fine.  The only thing that isn't working is where other apps need to establish a "connection" (i.e. SongBook Pro for example) then my app does not show up as a connectable MIDI app (or endpoint).  So I was wondering if there is a parameter or something I can do to make my app's midi show up.

**Reply by orchetect** [reply]
The `allEndpoints` broadcast mode is mostly a convenience. Again, I'm not totally sure you'd want to implement that in a public release but I think some apps do so I'll leave that to your discretion.

For apps that want to connect to yours, you would need to provide virtual input and output endpoints.

`addInput` and `addOutput` are what you want. Check out the documentation about creating endpoints - it's close to where creating connections docs are.

Then when you send messages, you would send using both the connection and the output at the same time.

**Reply by SNewtown** [reply]
Thanks for pointing me in the right direction, AddInput totally worked!

Only thing I'm not 100% on is the:

uniqueID: MIDIIdentifierPersistence

I tried with  MIDIIdentifierPersistence.adHoc and that seem to work, but should I be using a dedicated uniqueID so that it persists between sessions?

Thanks again!

**Reply by SNewtown** [reply]
Just noticed the option MIDIIdentifierPersistence.userDefaultsManaged

Which I can deduct stores the ID in user defaults so that should work to auto generate the ID and persist it?

**Reply by orchetect** [reply]
Yes the UserDefaults option is recommended and designed to make it easy. Just give it a unique key name it can use to store its data.
